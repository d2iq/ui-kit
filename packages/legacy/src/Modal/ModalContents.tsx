// tslint:disable
import classNames from "classnames/dedupe";
// tslint:enable
import GeminiScrollbar from "react-gemini-scrollbar";
import React from "react";
/**
 * Lifecycle of a Modal:
 * initial page load -> empty TransitionGroup
 * interaction changes open to true -> render modal content without scrollbars
 * get height of content -> rerender modal content and cap the height
 */

import { CSSTransition } from "react-transition-group";

import DOMUtil from "../Util/DOMUtil";
import { ModalProps, ModalState } from "./Modal";

// This value is used to designate "off-limits" vertical space, so that the
// modal never comes into contact with the edge of the viewport.
const MODAL_VERTICAL_INSET_DISTANCE = 48;

const METHODS_TO_BIND = [
  "calculateContentHeight",
  "closeModal",
  "handleBackdropClick",
  "handleKeyDown",
  "handleWindowResize"
];

export default class ModalContents extends React.Component<
  ModalProps,
  ModalState
> {
  public static defaultProps: Partial<ModalProps> = {
    closeByBackdropClick: true,
    transitionNameModal: "modal",
    transitionAppearTimeoutModal: 300,
    transitionEnterTimeoutModal: 300,
    transitionExitTimeoutModal: 300,
    transitionAppear: true,
    transitionEnter: true,
    transitionExit: true,
    useGemini: true,

    // Default classes.
    backdropClass: "modal-backdrop",
    bodyClass: "modal-body-wrapper",
    closeButtonClass: "modal-close",
    footerClass: "modal-footer",
    headerClass: "modal-header",
    modalClass: "modal modal-large",
    scrollContainerClass: "modal-body"
  };

  public lastConstrainedHeight;
  public lastViewportHeight;
  private footerRef = React.createRef<any>();
  private geminiRef = React.createRef<any>();
  private headerRef = React.createRef<any>();
  private innerContentContainerRef = React.createRef<any>();
  private innerContentRef = React.createRef<any>();
  private modalRef = React.createRef<any>();

  constructor(props) {
    super(props);

    this.state = {
      height: null
    };

    METHODS_TO_BIND.forEach(method => {
      this[method] = this[method].bind(this);
    });
  }

  componentDidUpdate(prevProps: ModalProps) {
    // If we don't already know the height of the content, we calculate it.
    if (this.props.open) {
      this.lastViewportHeight = Math.ceil(DOMUtil.getViewportHeight());
      window.requestAnimationFrame(this.calculateContentHeight);
    }

    if (this.props.open !== prevProps.open) {
      document.body.classList.toggle("no-overflow");
    }
  }

  UNSAFE_componentWillUpdate(nextProps) {
    // Reset the height of the content to null when the modal is closing so
    // that the height will be recalculated next time it opens.
    if (this.props.open && !nextProps.open) {
      this.setState({ height: null });
      this.removeKeydownListener();
    }

    if (!this.props.open && nextProps.open) {
      this.addKeydownListener();
    }
  }

  UNSAFE_componentWillMount() {
    if (this.props.open) {
      document.body.classList.add("no-overflow");
    }

    window.addEventListener("resize", this.handleWindowResize);
  }

  componentWillUnmount() {
    document.body.classList.remove("no-overflow");
    window.removeEventListener("resize", this.handleWindowResize);
  }

  handleBackdropClick() {
    if (this.props.closeByBackdropClick) {
      this.closeModal();
    }
  }

  handleKeyDown(event) {
    if (event.keyCode === 27) {
      this.closeModal();
    }
  }

  handleWindowResize() {
    // Return early if the modal is closed or not using Gemini.
    if (!this.props.open) {
      return;
    }

    const viewportHeight = Math.ceil(DOMUtil.getViewportHeight());

    // If the height of the viewport is getting shorter, or if it's growing
    // while the height is currently constrained, then we reset the restrained
    // height to null which will cause the height to be recalculated on the
    // next render.
    if (
      viewportHeight < this.lastViewportHeight ||
      (viewportHeight > this.lastViewportHeight && this.state.height !== null)
    ) {
      this.setState({ height: null });
    }

    this.lastViewportHeight = viewportHeight;
    window.requestAnimationFrame(this.calculateContentHeight);
  }

  addKeydownListener() {
    document.body.addEventListener("keydown", this.handleKeyDown);
  }

  removeKeydownListener() {
    document.body.removeEventListener("keydown", this.handleKeyDown);
  }

  calculateContentHeight() {
    // A full screen modal doesn't need to restrict its height.
    if (this.props.isFullScreen) {
      return;
    }

    let headerHeight = 0;
    let footerHeight = 0;
    let modalHeight = 0;
    let innerContentHeight = 0;

    if (this.headerRef != null && this.headerRef.current != null) {
      headerHeight = Math.ceil(
        this.headerRef.current.getBoundingClientRect().height
      );
    }

    if (this.footerRef != null && this.footerRef.current != null) {
      footerHeight = Math.ceil(
        this.footerRef.current.getBoundingClientRect().height
      );
    }

    if (this.modalRef != null && this.modalRef.current != null) {
      modalHeight = Math.ceil(
        this.modalRef.current.getBoundingClientRect().height
      );
    }

    if (this.innerContentRef != null && this.innerContentRef.current != null) {
      innerContentHeight = Math.ceil(
        this.innerContentRef.current.getBoundingClientRect().height
      );
    }

    const maxModalHeight =
      this.lastViewportHeight - MODAL_VERTICAL_INSET_DISTANCE;

    const totalModalContentHeight =
      innerContentHeight + headerHeight + footerHeight;

    // When the modal's content fits on the screen, both the modal and body
    // height can be set to `auto` (default).
    let nextInnerContentContainerHeight = "auto";
    let nextModalHeight = "auto";

    // When the modal's content is too large to fit on the screen, then we need
    // to explicitly set the body's height to its exact pixel value and the
    // modal's height to `100%`.
    const shouldConstrainHeight =
      totalModalContentHeight >= maxModalHeight ||
      this.lastViewportHeight < this.lastConstrainedHeight;

    if (shouldConstrainHeight) {
      const availableContentHeight = modalHeight - headerHeight - footerHeight;
      nextInnerContentContainerHeight = `${availableContentHeight}px`;
      nextModalHeight = "100%";

      // We need to keep track of the largest viewport height that results in a
      // constrained modal.
      if (
        this.lastConstrainedHeight == null ||
        this.lastViewportHeight > this.lastConstrainedHeight
      ) {
        this.lastConstrainedHeight = this.lastViewportHeight;
      }

      if (
        this.props.useGemini &&
        this.state.height !== availableContentHeight
      ) {
        this.setState({ height: availableContentHeight });
      }
    }

    if (
      this.innerContentContainerRef != null &&
      this.innerContentContainerRef.current != null
    ) {
      this.innerContentContainerRef.current.style.height = nextInnerContentContainerHeight;
    }
    if (this.modalRef != null && this.modalRef.current != null) {
      this.modalRef.current.style.height = nextModalHeight;
    }

    this.triggerGeminiUpdate();
  }

  closeModal() {
    this.props.onClose();
  }

  getCloseButton() {
    if (this.props.closeButton) {
      return this.props.closeButton;
    }

    return null;
  }

  getHeader() {
    if (this.props.showHeader === false) {
      return null;
    }

    return (
      <div className={this.props.headerClass} ref={this.headerRef}>
        {this.props.header}
        {this.props.subHeader}
      </div>
    );
  }

  getFooter() {
    if (this.props.showFooter === false) {
      return null;
    }

    return (
      <div className={this.props.footerClass} ref={this.footerRef}>
        {this.props.footer}
      </div>
    );
  }

  getModalContent() {
    const modalContent = (
      <div
        className={this.props.scrollContainerClass}
        ref={this.innerContentRef}
      >
        {this.props.children}
      </div>
    );

    // If the consume disables gemini or we don't know the height, then we
    // don't render with Gemini, unless the consumer is using specifying a
    // custom height.
    if (
      (!this.props.useGemini || this.state.height == null) &&
      this.props.modalHeight == null
    ) {
      return modalContent;
    }

    const geminiClasses = classNames(
      "container-scrollable",
      this.props.geminiClass
    );
    const geminiContainerStyle = { height: this.state.height };

    if (this.props.modalHeight) {
      geminiContainerStyle.height = this.props.modalHeight;
    }

    return (
      <GeminiScrollbar
        autoshow={false}
        className={geminiClasses}
        ref={this.geminiRef}
        style={geminiContainerStyle}
      >
        {modalContent}
      </GeminiScrollbar>
    );
  }

  getModal() {
    const modalBodyStyle: any = {};

    if (!this.props.open) {
      return null;
    }

    if (this.state.height != null) {
      modalBodyStyle.height = this.state.height;
    }

    return (
      <div ref={this.modalRef} className={this.props.modalClass}>
        {this.getCloseButton()}
        {this.getHeader()}
        <div
          className={this.props.bodyClass}
          style={modalBodyStyle}
          ref={this.innerContentContainerRef}
        >
          {this.getModalContent()}
        </div>
        {this.getFooter()}
      </div>
    );
  }

  getBackdrop() {
    if (!this.props.open) {
      return null;
    }

    return (
      // tslint:disable react-a11y-event-has-role
      <div
        className={this.props.backdropClass}
        onClick={this.handleBackdropClick}
      />
      // tslint:enable
    );
  }

  triggerGeminiUpdate() {
    if (
      this.geminiRef != null &&
      this.geminiRef.current != null &&
      this.geminiRef.current.scrollbar != null
    ) {
      this.geminiRef.current.scrollbar.update();
    }
  }

  render() {
    let modalContent: any = null;

    if (this.props.open) {
      modalContent = (
        <div
          className={classNames("modal-wrapper", this.props.modalWrapperClass)}
        >
          {this.getBackdrop()}
          {this.getModal()}
        </div>
      );
    }

    return (
      <CSSTransition
        in={this.props.open}
        appear={this.props.transitionAppear}
        enter={this.props.transitionEnter}
        exit={this.props.transitionExit}
        classNames={this.props.transitionNameModal}
        timeout={{
          enter: this.props.transitionEnterTimeout,
          exit: this.props.transitionExitTimeout,
          appear: this.props.transitionAppearTimeoutModal
        }}
      >
        <div>{modalContent}</div>
      </CSSTransition>
    );
  }
}
