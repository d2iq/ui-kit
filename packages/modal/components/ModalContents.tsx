import React from "react";
import { cx } from "@emotion/css";
import { modalWrapper } from "../style";
import { flex } from "../../shared/styles/styleUtils";
export interface ModalContentsProps {
  isOpen: boolean;
  onClose: () => void;
  ["data-cy"]?: string;
}

const ModalContents: React.FC<ModalContentsProps> = ({
  children,
  "data-cy": dataCy
}) => (
  <div
    className={cx(modalWrapper, flex({ direction: "column" }))}
    data-cy={dataCy}
  >
    {children}
  </div>
);

export default ModalContents;
