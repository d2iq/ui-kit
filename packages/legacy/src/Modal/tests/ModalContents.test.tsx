import React from "react";
import { mount } from "enzyme";
import * as emotion from "emotion";
import { createSerializer } from "jest-emotion";
import toJson from "enzyme-to-json";
import ModalContents from "../ModalContents";

expect.addSnapshotSerializer(createSerializer(emotion));

describe("ModalContents", () => {
  it("renders", () => {
    const onClose = jest.fn();
    const component = mount(
      <ModalContents open={true} onClose={onClose}>
        content
      </ModalContents>
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it("renders with a footer", () => {
    const onClose = jest.fn();
    const component = mount(
      <ModalContents open={true} onClose={onClose} showFooter={true}>
        content
      </ModalContents>
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it("renders with a close button", () => {
    const onClose = jest.fn();
    const component = mount(
      <ModalContents
        open={true}
        onClose={onClose}
        closeButton={<button>Close</button>}
      >
        content
      </ModalContents>
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  describe("#onClose", () => {
    it("should call onClose when the modal closes", () => {
      const onClose = jest.fn();
      const component = mount(
        <ModalContents open={true} onClose={onClose}>
          content
        </ModalContents>
      );

      expect(onClose).not.toHaveBeenCalled();
      component.find(".modal-backdrop").simulate("click");
      expect(onClose).toHaveBeenCalled();
    });

    it("does not call onClose if closeByBackdropClick is false", () => {
      const onClose = jest.fn();
      const component = mount(
        <ModalContents
          open={true}
          onClose={onClose}
          closeByBackdropClick={false}
        >
          content
        </ModalContents>
      );

      component.find(".modal-backdrop").simulate("click");
      expect(onClose).not.toHaveBeenCalled();
    });
  });

  describe("#handleWindowResize", () => {
    beforeEach(() => {
      jest
        .spyOn(window, "requestAnimationFrame")
        .mockImplementation(cb => cb());
    });

    it("should store the last viewport height on the instance", () => {
      const onClose = jest.fn();
      const component = mount(
        <ModalContents open={true} onClose={onClose}>
          content
        </ModalContents>
      );

      const instance = component.instance() as ModalContents;
      Object.defineProperty(window, "innerHeight", {
        writable: true,
        configurable: true,
        value: 500
      });
      window.dispatchEvent(new Event("resize"));

      expect(instance.lastViewportHeight).toEqual(500);
    });

    it("should reset the height stored in state to null when the viewport height is shrinking", () => {
      const onClose = jest.fn();
      const component = mount(
        <ModalContents open={true} onClose={onClose}>
          content
        </ModalContents>
      );
      const instance = component.instance() as ModalContents;

      instance.setState({ height: 300 });

      expect(component.state("height")).toBe(300);

      Object.defineProperty(window, "innerHeight", {
        writable: true,
        configurable: true,
        value: 500
      });
      window.dispatchEvent(new Event("resize"));

      Object.defineProperty(window, "innerHeight", {
        writable: true,
        configurable: true,
        value: 400
      });
      window.dispatchEvent(new Event("resize"));

      expect(component.state("height")).toBe(null);
    });
  });

  describe("overflow hidden on body", () => {
    it("should toggle no-overflow on body when updated to open", () => {
      const onClose = jest.fn();
      const component = mount(
        <ModalContents open={true} onClose={onClose}>
          content
        </ModalContents>
      );
      const classToggleSpy = spyOn(document.body.classList, "toggle");
      component.setProps({ open: false });

      expect(classToggleSpy).toHaveBeenCalledWith("no-overflow");
    });

    it("should toggle no-overflow on body when updated to close", () => {
      const onClose = jest.fn();
      const component = mount(
        <ModalContents open={false} onClose={onClose}>
          content
        </ModalContents>
      );
      const classToggleSpy = spyOn(document.body.classList, "toggle");
      component.setProps({ open: true });

      expect(classToggleSpy).toHaveBeenCalledWith("no-overflow");
    });

    it("should not change class on body when mounted close", () => {
      const onClose = jest.fn();
      const component = mount(
        <ModalContents open={false} onClose={onClose}>
          content
        </ModalContents>
      );
      const classAddSpy = spyOn(document.body.classList, "add");
      component.mount();

      expect(classAddSpy).not.toHaveBeenCalledWith("no-overflow");
    });

    it("should remove no-overflow on body when unmounted", () => {
      const onClose = jest.fn();
      const component = mount(
        <ModalContents open={true} onClose={onClose}>
          content
        </ModalContents>
      );
      const classRemoveSpy = spyOn(document.body.classList, "remove");
      component.unmount();

      expect(classRemoveSpy).toHaveBeenCalledWith("no-overflow");
    });
  });
});
