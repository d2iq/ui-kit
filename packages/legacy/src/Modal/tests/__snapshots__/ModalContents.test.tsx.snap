// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalContents renders 1`] = `
<ModalContents
  backdropClass="modal-backdrop"
  bodyClass="modal-body-wrapper"
  closeButtonClass="modal-close"
  closeByBackdropClick={true}
  footerClass="modal-footer"
  headerClass="modal-header"
  modalClass="modal modal-large"
  onClose={[MockFunction]}
  open={true}
  scrollContainerClass="modal-body"
  transitionAppear={true}
  transitionAppearTimeoutModal={300}
  transitionEnter={true}
  transitionEnterTimeoutModal={300}
  transitionExit={true}
  transitionExitTimeoutModal={300}
  transitionNameModal="modal"
  useGemini={true}
>
  <CSSTransition
    appear={true}
    classNames="modal"
    enter={true}
    exit={true}
    in={true}
    timeout={
      Object {
        "appear": 300,
        "enter": undefined,
        "exit": undefined,
      }
    }
  >
    <Transition
      appear={true}
      enter={true}
      exit={true}
      in={true}
      mountOnEnter={false}
      onEnter={[Function]}
      onEntered={[Function]}
      onEntering={[Function]}
      onExit={[Function]}
      onExited={[Function]}
      onExiting={[Function]}
      timeout={
        Object {
          "appear": 300,
          "enter": undefined,
          "exit": undefined,
        }
      }
      unmountOnExit={false}
    >
      <div>
        <div
          className="modal-wrapper"
        >
          <div
            className="modal-backdrop"
            onClick={[Function]}
          />
          <div
            className="modal modal-large"
          >
            <div
              className="modal-header"
            />
            <div
              className="modal-body-wrapper"
              style={Object {}}
            >
              <div
                className="modal-body"
              >
                content
              </div>
            </div>
            <div
              className="modal-footer"
            />
          </div>
        </div>
      </div>
    </Transition>
  </CSSTransition>
</ModalContents>
`;

exports[`ModalContents renders with a close button 1`] = `
<ModalContents
  backdropClass="modal-backdrop"
  bodyClass="modal-body-wrapper"
  closeButton={
    <button>
      Close
    </button>
  }
  closeButtonClass="modal-close"
  closeByBackdropClick={true}
  footerClass="modal-footer"
  headerClass="modal-header"
  modalClass="modal modal-large"
  onClose={[MockFunction]}
  open={true}
  scrollContainerClass="modal-body"
  transitionAppear={true}
  transitionAppearTimeoutModal={300}
  transitionEnter={true}
  transitionEnterTimeoutModal={300}
  transitionExit={true}
  transitionExitTimeoutModal={300}
  transitionNameModal="modal"
  useGemini={true}
>
  <CSSTransition
    appear={true}
    classNames="modal"
    enter={true}
    exit={true}
    in={true}
    timeout={
      Object {
        "appear": 300,
        "enter": undefined,
        "exit": undefined,
      }
    }
  >
    <Transition
      appear={true}
      enter={true}
      exit={true}
      in={true}
      mountOnEnter={false}
      onEnter={[Function]}
      onEntered={[Function]}
      onEntering={[Function]}
      onExit={[Function]}
      onExited={[Function]}
      onExiting={[Function]}
      timeout={
        Object {
          "appear": 300,
          "enter": undefined,
          "exit": undefined,
        }
      }
      unmountOnExit={false}
    >
      <div>
        <div
          className="modal-wrapper"
        >
          <div
            className="modal-backdrop"
            onClick={[Function]}
          />
          <div
            className="modal modal-large"
          >
            <button>
              Close
            </button>
            <div
              className="modal-header"
            />
            <div
              className="modal-body-wrapper"
              style={Object {}}
            >
              <div
                className="modal-body"
              >
                content
              </div>
            </div>
            <div
              className="modal-footer"
            />
          </div>
        </div>
      </div>
    </Transition>
  </CSSTransition>
</ModalContents>
`;

exports[`ModalContents renders with a footer 1`] = `
<ModalContents
  backdropClass="modal-backdrop"
  bodyClass="modal-body-wrapper"
  closeButtonClass="modal-close"
  closeByBackdropClick={true}
  footerClass="modal-footer"
  headerClass="modal-header"
  modalClass="modal modal-large"
  onClose={[MockFunction]}
  open={true}
  scrollContainerClass="modal-body"
  showFooter={true}
  transitionAppear={true}
  transitionAppearTimeoutModal={300}
  transitionEnter={true}
  transitionEnterTimeoutModal={300}
  transitionExit={true}
  transitionExitTimeoutModal={300}
  transitionNameModal="modal"
  useGemini={true}
>
  <CSSTransition
    appear={true}
    classNames="modal"
    enter={true}
    exit={true}
    in={true}
    timeout={
      Object {
        "appear": 300,
        "enter": undefined,
        "exit": undefined,
      }
    }
  >
    <Transition
      appear={true}
      enter={true}
      exit={true}
      in={true}
      mountOnEnter={false}
      onEnter={[Function]}
      onEntered={[Function]}
      onEntering={[Function]}
      onExit={[Function]}
      onExited={[Function]}
      onExiting={[Function]}
      timeout={
        Object {
          "appear": 300,
          "enter": undefined,
          "exit": undefined,
        }
      }
      unmountOnExit={false}
    >
      <div>
        <div
          className="modal-wrapper"
        >
          <div
            className="modal-backdrop"
            onClick={[Function]}
          />
          <div
            className="modal modal-large"
          >
            <div
              className="modal-header"
            />
            <div
              className="modal-body-wrapper"
              style={Object {}}
            >
              <div
                className="modal-body"
              >
                content
              </div>
            </div>
            <div
              className="modal-footer"
            />
          </div>
        </div>
      </div>
    </Transition>
  </CSSTransition>
</ModalContents>
`;
